### DISCUSSION TITLE
Pimcore-X Get file system path of an Asset

### DISCUSSION THREAD

**Discussion by maximilianherrmannblackbit** [pimcore/pimcore#11001] [Q&A / Support] [answered]
Hello everyone, I hope someone can help me.

I am in the process of upgrading a project to Pimcore-X. In the old version (6.x) there was a function to get the filesystem path of an asset:
`/** @var $asset Pimcore/Model/Asset */`
`$asset->getFileSystemPath();`.

This function has been removed in Pimcore-X.

Now my question: how can I get the file system path of a specific asset?
It must work even if the asset is saved in AWS S3 or some other storage-service.

Thank you very much for your help.

**Comment by NiklasBr** [comment, marked as answer]
You need the FlySystem storage engine now due to abstraction.
```php
$storage = Storage::get('asset');
if ($storage->fileExists($asset->getFullPath())) { /* do stuff */ }
```

**Reply by maximilianherrmannblackbit** [reply]
Thank you for you answer.

If i understand your example correctly, it just check if the file exists. But i do not get the file system path of the asset (something like "/var/www/path/to/asset/file.pdf"), right?

In Pimcore 5.8 i could do `$path = $asset->getFileSystemPath();` and know i need a alternativ function for this in Pimcore-X.

**Reply by NiklasBr** [reply]
It is no longer feasible to keep working with full and static system file paths starting with Pimcore X (due to support for so many different storage engines). Once you have the storage object and the asset object you have everything you need to move it, get its data, delete it, and so on...

The storage engine knows the base path (if one even exists), your asset knows its location in relation to its storage engine.

https://flysystem.thephpleague.com/v2/docs/

**Reply by maximilianherrmannblackbit** [reply]
Thank you for your explanation. Then i have to do it this way.

**Reply by kaurov** [reply]
Storage::class is marked as @internal in Pimcore X

See usage sample:
https://pimcore.com/docs/pimcore/10.2/Development_Documentation/Assets/Restricting_Public_Asset_Access.html

 $stream = $asset->getStream();
            return new StreamedResponse(function () use ($stream) {
                fpassthru($stream);
            }, 200, [
                'Content-Type' => 'application/pdf',
            ]);

**Comment by kovinet** [comment]
How about if I need filesize of an image thumbnail for example?  

The only method on 'Pimcore\Model\Asset\Image\Thumbnail' available is now `getPath()` which returns only relative path. So if I feed it to Storage with `$storage->fileSize($thumbnail->getThumbnail('contentimages')->getPath())` it doesn't work.

**Reply by NiklasBr** [reply]
Have you tried the following?
```php
/** @var Pimcore\Model\Asset\Image $asset */
$asset->getThumbnail('contentimages')->getFileSize();
```

**Reply by kovinet** [reply]
Yes, Thumbnail class doesn't have this method.

**Reply by kovinet** [reply]
There is a configured Storage for thumbnails, I am using the correct one: `$storage = Storage::get('thumbnail')` but for some reason I still get the error "Unable to retrieve the file_size for file at location"

**Reply by kovinet** [reply]
Okay I see now, the problem is that thumbnail is not yet generated, this only works for already generated thumbs but of course it won't trigger the thumbnail generation...
